go/vt/vtgate/planbuilder/abstract/filter.go

```diff
@@ -48,8 +48,20 @@ func (f *Filter) CheckValid() error {
 
 // PushPredicate implements the LogicalOperator interface
 func (f *Filter) PushPredicate(expr sqlparser.Expr, semTable *semantics.SemTable) (LogicalOperator, error) {
-	f.Predicates = append(f.Predicates, expr)
-	return f, nil
+	op, err := f.Source.PushPredicate(expr, semTable)
+	if err != nil {
+		return nil, err
+	}
+
+	if filter, isFilter := op.(*Filter); isFilter {
+		filter.Predicates = append(f.Predicates, filter.Predicates...)
+		return filter, err
+	}
+
+	return &Filter{
+		Source:     op,
+		Predicates: f.Predicates,
+	}, nil
```

Comment:
I guess this could also re-use `f` here? But I wasn't sure about the supposed semantics around `PushPredicate` and whether structs returned here are supposed to be modified / reused.

```suggestion
        f.Source = op
        return f, nil
```

Reply:
I think returning `f` does not make a strong difference.